### DISCUSSION TITLE
create pdf from a list using handelbars

### DISCUSSION THREAD

**Discussion by sergedahdouh** [AntlerVC/firetable#847] [Support / Q&A] [answered]
hello 

I have the following script 

`const derivative:Derivative = async ({row,ref,db,storage,auth})=>{
    const pdfFrom = require("pdf-from");
    const Handlebars = require("handlebars");

    const htmlString = `
    <!DOCTYPE html>

<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    
    
    
    <title>Invoice - Bootdey.com</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<script src="./Invoice Template_files/jquery-1.10.2.min.js.download"></script>
    <link href="./Invoice Template_files/bootstrap.min.css" rel="stylesheet">
	<script src="./Invoice Template_files/bootstrap.min.js.download"></script>
<style type="text/css">
@font-face {
  font-weight: 400;
  font-style:  normal;
  font-family: 'Circular-Loom';

  src: url('https://cdn.loom.com/assets/fonts/circular/CircularXXWeb-Book-cd7d2bcec649b1243839a15d5eb8f0a3.woff2') format('woff2');
}

@font-face {
  font-weight: 500;
  font-style:  normal;
  font-family: 'Circular-Loom';

  src: url('https://cdn.loom.com/assets/fonts/circular/CircularXXWeb-Medium-d74eac43c78bd5852478998ce63dceb3.woff2') format('woff2');
}

@font-face {
  font-weight: 700;
  font-style:  normal;
  font-family: 'Circular-Loom';

  src: url('https://cdn.loom.com/assets/fonts/circular/CircularXXWeb-Bold-83b8ceaf77f49c7cffa44107561909e4.woff2') format('woff2');
}

@font-face {
  font-weight: 900;
  font-style:  normal;
  font-family: 'Circular-Loom';

  src: url('https://cdn.loom.com/assets/fonts/circular/CircularXXWeb-Black-bf067ecb8aa777ceb6df7d72226febca.woff2') format('woff2');
}</style></head>
<body _c_t_common="1" style="" data-new-gr-c-s-check-loaded="14.1080.0" data-gr-ext-installed="" _c_t_j1="1">
<div class="col-md-12">   
 <div class="row">
		
        <div class="receipt-main col-xs-10 col-sm-10 col-md-6 col-xs-offset-1 col-sm-offset-1 col-md-offset-3">
            <div class="row">
    			<div class="receipt-header">
					<div class="col-xs-6 col-sm-6 col-md-6">
						<div class="receipt-left">
							<img class="img-responsive" alt="{{groupName}}" src="./Invoice Template_files/{{groupLogoURL}}" style="width: 71px; border-radius: 43px;">
						</div>
					</div>
					<div class="col-xs-6 col-sm-6 col-md-6 text-right">
						<div class="receipt-right">
							<h5>{{groupName}}</h5>
							
						</div>
					</div>
				</div>
            </div>
			
			<div class="row">
				<div class="receipt-header receipt-header-mid">
					<div class="col-xs-8 col-sm-8 col-md-8 text-left">
						<div class="receipt-right">
							<h5>{{customer}}</h5>
							<p><b>Booking for : </b> {{participantName}}</p>
							<p><b>Booking Date :</b> {{bookingDate}}</p>
							<p><b>Payment Status :</b> {{paymentStatus}}</p>
							<p><b>Ticket Status :</b> {{ticketStatus}}</p>
						</div>
					</div>
					<div class="col-xs-4 col-sm-4 col-md-4">
						<div class="receipt-left">
							<h3>{{ticketRef}}</h3>
						</div>
					</div>
				</div>
            </div>
			
            <div>
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Description</th>
                            <th>Amount</th>
                        </tr>
                    </thead>
                    <tbody>
                    {{#repeater}}
                        <tr>
                            <td class="col-md-9">{{eventName}}</td>
                            <td class="col-md-3"><i class="fa fa-inr"></i>{{currencyCode}} {{amountTotal}}</td>
                        </tr>
                    {{/repeater}}
                        <tr>
                            <td class="col-md-9"><p><b>Event From :</b> {{eventFromDate}} {{eventFromTime}} : {{eventToDate}} {{eventToTime}}</p></td>
                            
                        </tr>
                        <tr>
                            <td class="col-md-9"><p><b>Event Till :</b> {{eventToDate}} {{eventToTime}}</p></td>
                          
                        </tr>
                        <tr>
                            <td class="text-right">
                            <p>
                                <strong>Total Amount: </strong>
                            </p>
                            <p>
                                <strong>Late Fees: </strong>
                            </p>
							<p>
                                <strong>Payable Amount: </strong>
                            </p>
							<p>
                                <strong>Balance Due: </strong>
                            </p>
							</td>
                            <td>
                            <p>
                                <strong><i class="fa fa-inr"></i>{{currencyCode}} {{amountTotal}}</strong>
                            </p>
                            <p>
                                <strong><i class="fa fa-inr"></i>{{currencyCode}} 0</strong>
                            </p>
							<p>
                                <strong><i class="fa fa-inr"></i>{{currencyCode}} {{amountTotal}}</strong>
                            </p>
							<p>
                                <strong><i class="fa fa-inr"></i>{{currencyCode}} 0</strong>
                            </p>
							</td>
                        </tr>
                        <tr>
                           
                            <td class="text-right"><h2><strong>Status: </strong></h2></td>
                            <td class="text-left text-danger"><h2><strong><i class="fa fa-inr"></i> {{paymentStatus}}</strong></h2></td>
                        </tr>
                    </tbody>
                </table>
            </div>
			
			<div class="row">
				<div class="receipt-header receipt-header-mid receipt-footer">
					<div class="col-xs-8 col-sm-8 col-md-8 text-left">
						<div class="receipt-right">
							
							<h5 style="color: rgb(140, 140, 140);">Thank fors your participation.!</h5>
						</div>
					</div>
					
				</div>
            </div>
			
        </div>    
	</div>
</div>

<style type="text/css">
body{
background:#eee;
margin-top:20px;
}
.text-danger strong {
        	color: #9f181c;
		}
		.receipt-main {
			background: #ffffff none repeat scroll 0 0;
			border-bottom: 12px solid #333333;
			border-top: 12px solid #9f181c;
			margin-top: 50px;
			margin-bottom: 50px;
			padding: 40px 30px !important;
			position: relative;
			box-shadow: 0 1px 21px #acacac;
			color: #333333;
			font-family: open sans;
		}
		.receipt-main p {
			color: #333333;
			font-family: open sans;
			line-height: 1.42857;
		}
		.receipt-footer h1 {
			font-size: 15px;
			font-weight: 400 !important;
			margin: 0 !important;
		}
		.receipt-main::after {
			background: #414143 none repeat scroll 0 0;
			content: "";
			height: 5px;
			left: 0;
			position: absolute;
			right: 0;
			top: -13px;
		}
		.receipt-main thead {
			background: #414143 none repeat scroll 0 0;
		}
		.receipt-main thead th {
			color:#fff;
		}
		.receipt-right h5 {
			font-size: 16px;
			font-weight: bold;
			margin: 0 0 7px 0;
		}
		.receipt-right p {
			font-size: 12px;
			margin: 0px;
		}
		.receipt-right p i {
			text-align: center;
			width: 18px;
		}
		.receipt-main td {
			padding: 9px 20px !important;
		}
		.receipt-main th {
			padding: 13px 20px !important;
		}
		.receipt-main td {
			font-size: 13px;
			font-weight: initial !important;
		}
		.receipt-main td p:last-child {
			margin: 0;
			padding: 0;
		}	
		.receipt-main td h2 {
			font-size: 20px;
			font-weight: 900;
			margin: 0;
			text-transform: uppercase;
		}
		.receipt-header-mid .receipt-left h1 {
			font-weight: 100;
			margin: 34px 0 0;
			text-align: right;
			text-transform: uppercase;
		}
		.receipt-header-mid {
			margin: 24px 0;
			overflow: hidden;
		}
		
		#container {
			background-color: #dcdcdc;
		}
</style>

<script type="text/javascript">

</script>

<div id="fwcsnackbar"></div><div style="display: none" class="ubey-RecordingScreen-count-down ubey-RecordingScreen-count-down-container">
    <style>
        .ubey-RecordingScreen-count-down-container {
            position: fixed;
            height: 100vh;
            width: 100vw;
            top: 0;
            left: 0;
            z-index: 9999999999999;
            background-color: rgba(0, 0, 0, 0.2);
        }

        .ubey-RecordingScreen-count-down-content {
            position: absolute;
            display: flex;
            top: 50%;
            left: 50%;
            justify-content: center;
            align-items: center;
            color: white;
            height: 15em;
            width: 15em;
            transform: translate(-50%, -100%);
            background-color: rgba(0, 0, 0, 0.6);
            border-radius: 50%;
        }

        #ubey-RecordingScreen-count-count {
            font-size: 14em;
            transform: translateY(-2%);
        }
    </style>
    <div class="ubey-RecordingScreen-count-down-content">
        <span id="ubey-RecordingScreen-count-count"></span>
    </div>
</div></body><loom-container id="lo-engage-ext-container"><loom-shadow classname="resolved"></loom-shadow></loom-container><loom-container id="lo-companion-container"><loom-shadow classname="resolved"></loom-shadow></loom-container><grammarly-desktop-integration data-grammarly-shadow-root="true"></grammarly-desktop-integration></html>
    `;
    const template = Handlebars.compile(htmlString);

    const data = {
        groupName : row.bookedByName,
        currencyCode : row.currencyCode,
        repeater : [{eventName :row.eventName}]

				// [{eventName :row.eventName.stringify= JSON.stringify(row.eventName)}]
    };

    const result = template(data);

    const pdfFile = await pdfFrom.html(result);
		
    
    const file = await rowy.storage.upload.data(pdfFile, {
        fileName: "receiptHadndlebars.pdf",
        folderPath: ref.id
      });
      return [file]
      }`

and row is json with data when the script is executed the pdf template return object.objject please i need some support 
what i am looking is the template should have multiple raws based on the row data 

<img width="1327" alt="Screenshot 2022-09-28 at 05 58 36" src="https://user-images.githubusercontent.com/40029858/192684492-ad364c45-e631-442a-a82c-14205f40f720.png">

**Comment by harinij** [comment, marked as answer]
Hi @sergedahdouh we are starting to do office hours now and in case you are still looking for a solution for this, I recommend attending one of the upcoming office hours so that we can go over this on that chat, sign up here: https://discord.gg/vZbFQzje?event=1059820599184195604. We have a tutorial on PDF generation here: https://docs.rowy.io/tutorials/pdf-generator
